### DISCUSSION TITLE
removing specifc fields from responses

### DISCUSSION THREAD

**Discussion by YakirOren** [pocketbase/pocketbase#5369] [Q&A]
Hi,
I have a user collection and each user has a role.
for users with weaker roles, I want to exclude certain fields from being sent.

I have seen this discussion https://github.com/pocketbase/pocketbase/discussions/4585
I'm considered creating a new collection, but also wanted to know if there's a way to do this in the backend.

currently I'm using this OnRecordViewRequest hook to remove the data from the field
```go
authRecord, ok := e.HttpContext.Get("authRecord").(*models.Record)
if !ok {
	return nil
}

if authRecord.Get("role") == "weak_role" {
	e.Record.Set("excluded", "")
}
```
I don't want the key to appear at all in the responses

Thanks

**Comment by ganigeorgiev** [comment]
> I don't want the key to appear at all in the responses

If it is part of the collection schema then it is not possible (_unless you modify `e.Record.Collection().Schema` but without details about your use case I don't recommend it as it is too error prone and it could affect other records from the same collection_).

A better approach would be to move your private fields in a separate collection that will have its own API rules and link back to the main one with a `relation` field.

_Side-note:_ With the next v0.23.0 release (_there are no ETAs yet_) there are 2 new futures that when combined could eventually make it easier:
- option to mark a field as "Hidden", aka. hide it from the API responses and filters (by default only admins/superusers will have access to the field and will be able to modify it)
- new `OnRecordEnrich` hook which is essentially triggered for each Record API response (_or more accurately when `apis.EnrichRecords` is called, which is true for every builtin Record related response_)

**Reply by kendallroth** [reply]
> option to mark a field as "Hidden", aka. hide it from the API responses and filters (by default only admins/superusers will have access to the field and will be able to modify it)

Does the mention of "by default" mean it will be possible to override whether a field is visible based on some custom criteria (ie. based on a field in the requesting user's record)?

**Reply by ganigeorgiev** [reply]
By "default" I meant that the hidden state of a field would be based on its `Hidden/Private` option (_I still can't decide on the name_) that will be available in the Admin UI. If a field is marked as "Hidden/Private" only admins/superusers will be able to read, filter or modify (_the only slight exception from this rule is for auth collections and the `password` field in order to allow users to set/change their password_).

Developers will have the option to additionally hide/skip fields from the serialization using the `record.Hide(fieldNames...)` method that you could eventually set in the `OnRecordEnrich` hook. The "hidden" state in this case is only valid for the duration of the specific request (_or more accurately for the lifetime of the record instance_). For example:

```go
app.OnRecordEnrich("yourCollection").BindFunc(func(e *core.RecordEnrichEvent) error {
    if e.RequestInfo.Auth != nil &&
        !e.RequestInfo.Auth.IsSuperuser() &&
        e.RequestInfo.Auth.GetString("role") == "weak_role" {
        e.Record.Hide("yourField") // hide from the serialization
    }

    return e.Next()
})
```

**Reply by c-nv-s** [reply]
> By "default" I meant that the hidden state of a field would be based on its `Hidden/Private` option (_I still can't decide on the name_) that will be available in the Admin UI. If a field is marked as "Hidden/Private" only admins/superusers will be able to read, filter or modify (_the only slight exception from this rule is for auth collections and the `password` field in order to allow users to set/change their password_).
> 
> Developers will have the option to additionally hide/skip fields from the serialization using the `record.Hide(fieldNames...)` method that you could eventually set in the `OnRecordEnrich` hook. The "hidden" state in this case is only valid for the duration of the specific request (_or more accurately for the lifetime of the record instance_). For example:
> 
> ```go
> app.OnRecordEnrich("yourCollection").BindFunc(func(e *core.RecordEnrichEvent) error {
>     if e.RequestInfo.Auth != nil &&
>         !e.RequestInfo.Auth.IsSuperuser() &&
>         e.RequestInfo.Auth.GetString("role") == "weak_role" {
>         e.Record.Hide("yourField") // hide from the serialization
>     }
> 
>     return e.Next()
> })
> ```

according to the docs for js if I wanted to hide ``collectionId`` or ``collectionName`` from being shown in the results should the following not work?

```
// pb_hooks/01.pb.js

onRecordEnrich((e) => {
   
        e.record.hide("collectionId")
 

    e.next()
}, "articles")
```

because it doesn't seem to work.

**Reply by ganigeorgiev** [reply]
@c-nv-s It works correctly for me in the sense that the above will hide the `collectionId` field from the "articles" record (to hide also the `collectionName` you have to explicitly type it).

If it doesn't work, then elaborate more on your setup and what PocketBase version you are using (this feature is available with v0.23+), but please don't hijack old discussions.

**Reply by c-nv-s** [reply]
> @c-nv-s It works correctly for me in the sense that the above will hide the `collectionId` field from the "articles" record (to hide also the `collectionName` you have to explicitly type it).
> 
> If it doesn't work, then elaborate more on your setup and what PocketBase version you are using (this feature is available with v0.23+), but please don't hijack old discussions.

sorry, turns out the problem was completely different (wrong client context), sorry for hijacking the thread

**Comment by axshor** [comment]
great! waittting this feature!
